relay-server/src/actors/spooler/mod.rs

```diff
@@ -684,6 +697,47 @@ impl OnDisk {
             relay_statsd::metric!(gauge(RelayGauges::BufferEnvelopesDiskCount) = *count);
         }
     }
+
+    fn extract_key(&self, row: SqliteRow) -> Result<QueueKey, BufferError> {
+        let own_key: &str = row.try_get("own_key").map_err(BufferError::FetchFailed)?;
+        let sampling_key: &str = row
+            .try_get("sampling_key")
+            .map_err(BufferError::FetchFailed)?;
+        let queue_key = QueueKey {
+            own_key: ProjectKey::parse(own_key).map_err(BufferError::ParseProjectKeyFailed)?,
+            sampling_key: ProjectKey::parse(sampling_key)
+                .map_err(BufferError::ParseProjectKeyFailed)?,
+        };
+        Ok(queue_key)
+    }
+
+    /// Returns the index from the on-disk spool.
+    async fn get_spooled_index(
+        &self,
+    ) -> Result<BTreeMap<ProjectKey, BTreeSet<QueueKey>>, BufferError> {
+        let keys = sql::get_keys()
+            .fetch_all(&self.db)
+            .await
+            .map_err(BufferError::FetchFailed)?;
+
+        let extracted_keys: Result<Vec<_>, BufferError> = keys
+            .into_iter()
+            // Collect only keys we could extract.
+            .map(|row| self.extract_key(row))
+            .collect();
+
+        // Fold the list into the index format.
+        let index = extracted_keys?.into_iter().fold(
+            BTreeMap::new(),
+            |mut acc: BTreeMap<ProjectKey, BTreeSet<QueueKey>>, key| {
+                acc.entry(key.own_key).or_default().insert(key);
+                acc.entry(key.sampling_key).or_default().insert(key);
+                acc
+            },
+        );
```

Comment:
I could do this in one go, without allocating a vec, but that means I will have to also ignore the errors here, and filter them out, and report them in `extract_key` instead.